src/tui.rs

```diff
@@ -278,9 +278,55 @@ impl Tui {
 			};
 
 			// here we calculate how many pages can fit in the available area.
+			// We do this in two passes so we don't need to collect the images into an
+			// intermediate vector: first pass sums widths to determine centering, second
+			// pass renders pages one-at-a-time. We still avoid keeping references across
+			// the two passes.
+			let mut total_width: u16 = 0;
+			let mut pages_shown: usize = 0;
+
+			for (idx, page) in self.rendered[self.page..].iter_mut().enumerate() {
+				// only consider pages that have an image ready
+				if page.img.is_none() {
+					break;
+				}
+
+				if let Some(max) = self.page_constraints.max_wide {
+					if idx >= max.get() {
+						break;
+					}
+				}
+
+				let width = page.img.as_mut().map(|img| img.w_h().0).unwrap_or(0);
+
+				// stop if this would overflow the available area
+				match total_width.checked_add(width) {
+					Some(new_total) if new_total <= img_area.width => {
+						total_width = new_total;
+						pages_shown += 1;
+					}
+					_ => break
+				}
+			}
+
+			if pages_shown == 0 {
+				// If none are ready to render, just show the loading thing
+				Self::render_loading_in(frame, img_area);
+				return KittyDisplay::ClearImages;
+			}
+
+			execute!(stdout(), BeginSynchronizedUpdate).unwrap();
+
+			self.last_render.pages_shown = pages_shown;
+
+			let unused_width = img_area.width - total_width;
+			self.last_render.unused_width = unused_width;
+			img_area.x += unused_width / 2;
+
+			// Collect mutable references to the images for rendering in a single pass so
+			// we satisfy the borrow checker, then render them below.
```

Comment:
Sorry, but I'm confused as to why this whole section is necessary - it seems to just be duplicating the work done below. Is there a reason that we need to do it twice, or is this doing something different?